src/python/pants/build_graph/build_graph.py

```diff
@@ -458,15 +505,34 @@ def transitive_subgraph_of_addresses_bfs(self, addresses, predicate=None, levele
       out of the closure.  If it is given, any Target which fails the predicate will not be
       walked, nor will its dependencies.  Thus predicate effectively trims out any subgraph
       that would only be reachable through Targets that fail the predicate.
-    :param function leveled_predicate: Behaves identically to predicate, but takes the depth of the
+    :param function dep_predicate: Takes two parameters, the current target and the dependency of
+      the current target. If this parameter is not given, no dependencies will be filtered
+      when traversing the closure. If it is given, when the predicate fails, the edge to the dependency
+      will not be expanded.
+    :param function leveled_predicate: Deprecated. Behaves identically to predicate, but takes the depth of the
       target in the search tree as a second parameter, and it is checked just before a dependency is
       expanded.
     """
+    deprecated_conditional(
+      lambda: leveled_predicate is not None,
+      '1.6.0.dev0',
+      'leveled_predicate',
+      '''
+      Deprecated property leveled_predicate used. Please migrate to using dep_predicate.
+      '''
+    )
+    if leveled_predicate and dep_predicate:
+      raise ValueError('Cannot specify both leveled_predicate and dep_predicate')
     ordered_closure = OrderedSet()
     # Use the DepthAgnosticWalk if we can, because DepthAwareWalk does a bit of extra work that can
     # slow things down by few millis.
-    walker = self.DepthAwareWalk if leveled_predicate else self.DepthAgnosticWalk
-    walk = walker()
+    walk = None
```

Comment:
The walk initialization code is duplicated between the two methods. Could either have a helper, or make a factory function on your walk base class.

Reply:
yep. Adding a factory method for it.